### PR TITLE
feat(video): add optional SAM 2 propagation backend

### PR DESCRIPTION
**Pull request by abhiksark**
## Summary
- add a lazy, Qt-free SAM 2 adapter behind the frozen propagation backend contract
- expose Auto, OpenCV, and explicit SAM 2 settings with local checkpoint/config paths
- keep OpenCV as the portable fallback and keep Torch/SAM 2 out of every published extra
- convert masks to track-typed polygons/boxes, preserve manual anchors, stream previews, and record occlusion gaps
- add settings/UI/import/selection/adapter coverage plus fixed Linux light/dark 1x/2x screenshots

## Qualification
- `1037 passed, 1 skipped`
- changed-file Ruff, `git diff --check`, Python 3.8 AST parsing, and compileall passed
- wheel/sdist build and Twine checks passed; wheel contains `libs/core/video_sam2.py` and no Torch/SAM 2 requirement
- explicit unavailable SAM 2 leaves the canonical model and undo history unchanged
- official-source CUDA smoke: SAM 2 commit `2b90b9f5ceec907a1c18123530e92e794ad901a4`, SAM 2.1 Hiera Tiny checkpoint SHA-256 `7402e0d864fa82708a20fbd15bc84245c2f26dff0eb43a4b5b93452deb34be69`, PyTorch 2.12.0+cu130, torchvision 0.27.0+cu130, RTX A6000
- two consecutive 24-frame propagation runs succeeded and returned PyTorch allocated/reserved CUDA memory to 0 after each teardown

## Notes
- no runtime or model download is performed
- explicit SAM 2 never silently falls back; Auto selects it only when all platform/runtime/file gates pass
- issue #99 remains open until the 3.5.0 public release and evidence are verified

## Summary by CodeRabbit

* **New Features**
  * Added optional SAM 2 support for whole-video mask propagation.
  * Added settings to select the propagation backend and provide SAM 2 model/configuration paths.
  * Automatic mode can fall back to the standard video propagation method when SAM 2 is unavailable.
  * Added cancellation, progress reporting, correction regeneration, and project-wide video undo behavior.

* **Documentation**
  * Updated setup guidance, requirements, fallback behavior, workflow, export behavior, and settings screenshots.
  * Clarified that SAM 2 dependencies and model files are not bundled.

### CODE REVIEW

File: docs/features/smart-video-annotation.md [hunk 1]
```diff
@@ -39,14 +39,16 @@ because variable-frame-rate media has no reliable integer frame index.
...
 
 ## Tracks, anchors, and interpolation
 
 Drawing a rectangle or polygon on a video frame creates a UUID-backed track and
-an accepted manual observation at the current PTS. The **Tracks** tab shows the
-track label, span, color, visibility, and pending-review count. Renaming a
+an accepted manual observation at the current PTS. The unified **Objects**
+inspector shows the track label, span, color, visibility, provenance, render
+state, and pending-review count, including tracks absent from the current
+frame. Renaming a
```
[the remark this answers is not in the document]
Comment (abhiksark) [reply docs/features/smart-video-annotation.md:51]: Updated the video quick start to use the unified Objects inspector, integrated timeline, and Tools menu; regenerated PKG-INFO from README.

File: docs/features/smart-video-annotation.md [hunk 2]
```diff
@@ -58,14 +60,61 @@ amber dashed shapes are pending tracker suggestions.
...
+that track; an absence anchor or failed track produces a stable ``occluded``,
+``low_confidence``, ``out_of_frame``, or ``scene_cut`` gap while other tracks
+continue. Cancel, document replacement, shutdown, stale state, or a backend or
+decoder failure clears the preview without changing the durable project.
+
+Editing tracker or interpolated geometry creates an accepted manual correction
+immediately. A separate background job regenerates only the open segments up
+to the nearest manual anchor on either side. Successful regeneration is a
+second undo entry; failure or stale state preserves the correction and all
+previous generated data.
```
[the remark this answers is not in the document]
Comment (abhiksark) [reply docs/features/smart-video-annotation.md:94]: Updated the top-level summary for multi-type whole-video propagation, preview-only processing, atomic commits, and the separate legacy suggestion flow; regenerated PKG-INFO.

File: libs/core/video_sam2.py [hunk 3]
```diff
@@ -0,0 +1,622 @@
...
+        try:
+            _config_name(config_path, sam2_module)
+        except (AttributeError, RuntimeError, ValueError) as exc:
+            reasons.append(str(exc))
+        try:
+            source_install = _source_install(metadata_lookup)
+        except (OSError, ValueError):
+            source_install = False
+        if not source_install:
+            reasons.append('install official SAM 2 from a source checkout')
```
[the remark this answers is not in the document]
Comment (abhiksark) [reply libs/core/video_sam2.py:129]: Fixed by treating TypeError as an availability failure and added a namespace-package regression test.

File: libs/core/video_sam2.py [hunk 4]
```diff
@@ -0,0 +1,622 @@
...
+            path = os.path.join(directory, '%08d.jpg' % len(points))
+            if not cv2.imwrite(path, cv2.cvtColor(
+                    array, cv2.COLOR_RGB2BGR),
+                    [int(cv2.IMWRITE_JPEG_QUALITY), 95]):
+                raise RuntimeError('failed to stage a video frame for SAM 2')
+            points.append(int(frame.pts))
+        if not points or request.current_pts not in points:
+            raise RuntimeError(
+                'current propagation frame is no longer decodable')
+        return tuple(points), int(width), int(height)
```
[the remark this answers is not in the document]
Comment (abhiksark) [reply libs/core/video_sam2.py:366]: Fixed by staging only current-to-endpoint frames for the requested direction. Revalidated real official-source SAM 2 on CUDA in both directions: 23 observations, zero gaps, and zero allocated/reserved bytes after each run.

File: libs/core/video_sam2.py [hunk 5]
```diff
@@ -0,0 +1,622 @@
...
+                            raw_mask = masks.get(str(track_id))
+                            if raw_mask is None:
+                                mark_gap(track_id, track_state, pts)
+                                continue
+                            mask, quality = _mask_data(raw_mask, np)
+                            geometry = _mask_geometry(
+                                mask, track_state['rectangle'], cv2, np)
+                            if geometry is None:
+                                mark_gap(track_id, track_state, pts)
+                                continue
```
[the remark this answers is not in the document]
Comment (abhiksark) [reply libs/core/video_sam2.py:550]: Not changed: the release contract deliberately requires a failed track to remain inactive until its next accepted manual anchor or the media boundary. This matches deterministic gap/barrier semantics across propagation backends; automatic SAM 2 reacquisition would violate that locked behavior.

File: libs/core/video_sam2.py [hunk 6]
```diff
@@ -0,0 +1,622 @@
...
+                    predictor = None
+                    results = None
+                    mask_logits = None
+                    masks = None
+                    raw_mask = None
+                    gc.collect()
+                    try:
+                        _release_cuda_memory(torch)
+                    except (AttributeError, RuntimeError):
+                        pass
```
[the remark this answers is not in the document]
Comment (abhiksark) [reply libs/core/video_sam2.py:617]: Fixed by guarding expected reset/state cleanup failures while retaining unconditional reference clearing, garbage collection, and CUDA cache cleanup, so cleanup cannot replace the original cancellation or backend error.

File: README.rst [hunk 7]
```diff
@@ -101,6 +101,17 @@ Workflow and Interface
     generated geometry creates a manual correction first, then regenerates only
     the bounded neighboring segments as a separate undoable change.
 
+**Optional SAM 2 Video Backend**
+    Whole-video propagation can use an official, source-installed SAM 2 on
+    Linux/CUDA while OpenCV remains the portable default and fallback. Open
```
[the remark this answers is not in the document]
Comment (abhiksark) [reply README.rst]: Updated README to state that Auto is the default selection and OpenCV is the portable fallback; regenerated PKG-INFO.
